src/Features/Core/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -268,6 +276,26 @@ protected OptionSet GetOptions(Document document, CompletionTriggerInfo triggerI
 
         protected abstract Task<AbstractSyntaxContext> CreateContext(Document document, int position, CancellationToken cancellationToken);
 
+        private async Task<AbstractSyntaxContext> GetOrCreateContext(Document document, int position, CancellationToken cancellationToken)
+        {
+            lock (cacheGate)
+            {
+                if (document == cachedDocument && position == cachedPosition)
+                {
+                    return cachedContext;
```

Comment:
Note that there's a Task<T> allocated here. I don't think it'll become a bottleneck but it could be eliminated by splitting this async method into a "non-async, but Task returning" method and a private async helper method. The non-async method checks the cache and, if there's a cache hit, the cachedContext (which is stored as a Task<AbstractSyntaxCOntext>) is returned. A cache miss calls the 2nd method.
(We use this pattern elsewhere and, now that I've written it up, I think it could be encapsulated in a helper class)